### PR TITLE
Add experimental artifacts API

### PR DESCRIPTION
**Pull request by serinamarie**
### Overview

Exposes Artifacts API (server-side) atop new `artifact` table behind a `PREFECT_EXPERIMENTAL_ENABLE_ARTIFACTS` flag. Cloud implementation: https://github.com/PrefectHQ/nebula/pull/3513

Information related to 
1. the data and schema migration
2. related work this builds off of 

can be found in https://github.com/PrefectHQ/prefect/pull/8164

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

File: src/prefect/server/models/artifacts.py [hunk 1]
```diff
@@ -48,3 +50,127 @@ async def read_artifact(
 
     result = await session.execute(query)
     return result.scalar()
+
+
+@inject_db
+async def _apply_artifact_filters(
```
Comment (zangell44) [inline src/prefect/server/models/artifacts.py:56]: Hmmm this doesn't quite look right to me.

Comment (zangell44) [reply src/prefect/server/models/artifacts.py:56]: Can we also add a test that uses filtering?

**Review by zangell44 — CHANGES_REQUESTED** [review]
Looks good other than the comment on applying filters!

File: src/prefect/server/models/artifacts.py (hunk 1 above)
Comment (serinamarie) [reply src/prefect/server/models/artifacts.py:56]: Yeah, it wasn't right, good catch!

Comment (serinamarie) [reply src/prefect/server/models/artifacts.py:56]: I'll need add this into the `nebula` as well

File: src/prefect/server/models/artifacts.py [hunk 2]
```diff
@@ -48,3 +50,127 @@ async def read_artifact(
...
+        exists_clause = select(db.FlowRun).where(
+            db.Artifact.flow_run_id == db.FlowRun.id
+        )
+
+        if flow_run_filter:
+            exists_clause = exists_clause.where(flow_run_filter.as_sql_filter(db))
+        if task_run_filter:
+            exists_clause = exists_clause.join(
+                db.TaskRun,
+                db.TaskRun.flow_run_id == db.FlowRun.id,
```
Comment (zangell44) [inline src/prefect/server/models/artifacts.py:77]: Should we be joining task run and artifact instead here?

Comment (serinamarie) [reply src/prefect/server/models/artifacts.py:77]: Ah, yes 👍

**Review by zangell44 — APPROVED** [review]
Excellent work!

**Pull request merged**